lib/signs/utilities/early_am_suppression.ex

```diff
@@ -0,0 +1,308 @@
+
+                paginate(
+                  top_content,
+                  # Make zone nil in order to prevent the usual paging platform message
+                  {%{bottom | zone: nil},
+                   %Message.PlatformPredictionBottom{
+                     stop_id: bottom.stop_id,
+                     minutes: bottom.minutes
+                   }}
+                )
+
+              true ->
+                {top, _} = top_content
+                {bottom, _} = bottom_content
+                {top, bottom}
+            end
+
+          match?({%Message.Predictions{}, _}, top_content) or
+              match?({%Message.StoppedTrain{}, _}, top_content) ->
+            {top, _} = top_content
+            {top, map_to_single_line_content(bottom_content)}
+
+          match?({%Message.Predictions{}, _}, bottom_content) or
+              match?({%Message.StoppedTrain{}, _}, bottom_content) ->
+            {bottom, _} = bottom_content
+
+            case map_to_single_line_content(top_content) do
+              %EarlyAm.DestinationScheduledTime{} = top ->
+                {bottom, top}
+
+              %Headways.Paging{} = top ->
+                {bottom, top}
+
+              top ->
+                {top, bottom}
+            end
+
+          match?({%Message.GenericPaging{}, _}, top_content) and
+              match?({%Message.GenericPaging{}, _}, bottom_content) ->
+            {top, _} = top_content
+            {bottom, _} = bottom_content
+            {top, bottom}
+
+          match?({{%Headways.Top{}, _}, {%Headways.Top{}, _}}, {top_content, bottom_content}) ->
+            routes =
+              Signs.Utilities.SourceConfig.sign_routes(sign.source_config)
+              |> PaEss.Utilities.get_unique_routes()
+
+            {t1, t2} = top_content
+            {%{t1 | routes: routes}, t2}
+
+          true ->
+            paginate(top_content, bottom_content)
+        end
+
+      status ->
+        get_early_am_content(
+          sign,
+          messages,
+          schedule,
+          status,
+          current_time
+        )
+    end
+  end
+
+  defp get_mezzanine_early_am_content(
+         messages,
+         sign,
+         schedule,
+         statuses,
+         current_time
+       ) do
+    {top_scheduled, bottom_scheduled} = schedule
+
+    {top_message, bottom_message} =
+      case messages do
+        # JFK/UMass case
+        {%Message.GenericPaging{messages: [prediction, headway_top]},
+         %Message.GenericPaging{messages: [_, headway_bottom]}} ->
+          # Unpack the generic paging message from the normal content generation
+          # - The headway top message will get filtered out and headways will be re-fetched
+          # - The prediction can now be filtered out or trigger the special case logic in
+          #   do_early_am_suppression since we reset the zone to "m" and the generic paging
+          #   will be reconstructed
+          {%Message.Headways.Paging{
```

Comment:
Did some testing and seems to work. I also added a couple unit test cases for the cases you called out.

While looking at this, I did find other JFK/UMass cases that are unhandled, but they're not directly related to Early AM. When we check whether or not the messages and related data need to be flipped, we don't consider the single-line paging alert messages. Granted, these messages seldom, if ever, get used since single-direction shuttles and closures don't really happen at the T. It's possible in to configure those alerts that as single-direction in Alerts UI though, so we should account for it. I don't want to throw that into this PR though since it's large and confusing enough already.

Reply:
Thanks, sounds reasonable to put off that case if we don't expect it to happen in practice.